log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java

```diff
@@ -0,0 +1,189 @@
+
+            boolean isExpungeStaleEntries = true;
+            for (int i = 0; i < numberOfLoggers; i++) {
+                if (loggerRefByName.containsKey(loggerNamePrefix + i)) {
+                    isExpungeStaleEntries = false;
+                    break;
+                }
+            }
+            assertTrue(
+                    isExpungeStaleEntries,
+                    "Stale WeakReference entries were not removed from the inner map for MessageFactory");
+        });
+    }
+
+    @Test
+    void testExpungeStaleMessageFactoryEntry() {
+        SimpleMessageFactory mockMessageFactory = new SimpleMessageFactory();
+        Logger logger = registry.computeIfAbsent(
+                "testLogger", mockMessageFactory, (name, factory) -> new MockLogger(loggerContext, name, factory));
+        logger.info("Using logger {}", logger.getName());
+        logger = null;
+
+        await().atMost(10, SECONDS).pollInterval(100, MILLISECONDS).untilAsserted(() -> {
+            System.gc();
+            System.runFinalization();
+            registry.getLogger("triggerExpunge", mockMessageFactory);
+
+            Map<MessageFactory, Map<String, WeakReference<Logger>>> loggerRefByNameByMessageFactory =
+                    reflectAndGetLoggerMapFromRegistry();
+            assertNull(
+                    loggerRefByNameByMessageFactory.get(mockMessageFactory),
+                    "Stale MessageFactory entry was not removed from the outer map");
+        });
+    }
+
+    @Test
+    void testConcurrentAccess() throws InterruptedException {
+        int threadCount = 10;
+        ExecutorService executor = Executors.newFixedThreadPool(threadCount);
+        CountDownLatch latch = new CountDownLatch(threadCount);
+
+        for (int i = 0; i < threadCount; i++) {
+            executor.submit(() -> {
+                registry.computeIfAbsent(
+                        "testLogger", messageFactory, (name, factory) -> new MockLogger(loggerContext, name, factory));
+                latch.countDown();
+            });
+        }
+
+        latch.await();
+        executor.shutdown();
+
+        // Verify logger was created and is accessible after concurrent creation
+        assertNotNull(
+                registry.getLogger("testLogger", messageFactory),
+                "Logger should be accessible after concurrent creation");
+    }
+
+    private Map<MessageFactory, Map<String, WeakReference<Logger>>> reflectAndGetLoggerMapFromRegistry()
+            throws NoSuchFieldException, IllegalAccessException {
+        Field loggerMapField = registry.getClass().getDeclaredField("loggerRefByNameByMessageFactory");
+        loggerMapField.setAccessible(true);
+        @SuppressWarnings("unchecked")
+        Map<MessageFactory, Map<String, WeakReference<Logger>>> loggerMap =
+                (Map<MessageFactory, Map<String, WeakReference<Logger>>>) loggerMapField.get(registry);
+        return loggerMap;
+    }
+
+    private class MockLogger extends Logger {
```

Comment:
I don't think you really need this definition. All your `new MockLogger(...)` calls can be replaced with `new Logger(...) {}` instead.